### PR TITLE
ARROW-10173: [Rust][DataFusion] Implement support for direct comparison to scalar values

### PR DESCRIPTION
**Pull request by yordan-pavlov**
This PR addresses the inefficient comparison to scalar values, where an array is built with the scalar value repeated, by 
changing the return value of expressions from `Result<ArrayRef>` to `Result<ColumnarValue>`  where `ColumnarValue` is defined as:
```
pub enum ColumnarValue {
    /// Array of values
    Array(ArrayRef),
    /// A single value 
    Scalar(ScalarValue)
}
```

This enables scalar values to be used in comparison operators directly, and for the simple query used in the benchmark ("select f32, f64 from t where f32 >= 250 and f64 > 250") shows approximately 10x performance improvement:

before:
filter_scalar time: [35.733 ms 36.613 ms 37.924 ms]

after:
filter_scalar time: [3.5938 ms 3.6450 ms 3.7035 ms]
change: [-90.048% -89.846% -89.625%] (p = 0.00 < 0.05)

I have also added a benchmark to compare the change in performance when comparing two arrays (using query "select f32, f64 from t where f32 >= f64") and it is negligible:

before:
filter_array time: [11.601 ms 11.656 ms 11.718 ms]

after:
filter_array time: [11.854 ms 11.957 ms 12.070 ms]
change: [+1.8032% +3.6391% +5.5671%] (p = 0.00 < 0.05)

@andygrove @alamb let me know what you think

### CODE REVIEW

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (andygrove) [inline rust/datafusion/src/physical_plan/expressions.rs:1404]: It looks like this block is duplicated for the two match arms and could be moved into a function

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (andygrove) [inline rust/datafusion/src/physical_plan/expressions.rs:1007]: Could you add a comment explaining what is happening here? I am not familiar with this syntax

File: rust/datafusion/src/physical_plan/mod.rs
Comment (andygrove) [inline rust/datafusion/src/physical_plan/mod.rs:104]: Perhaps we could push the new `ColumnarValue` enum down to the core arrow crate since it isn't specific to DataFusion?

**Comment by andygrove** [comment]
@yordan-pavlov I took a quick skim through and this is looking really good! Could you rebase?

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (yordan-pavlov) [reply rust/datafusion/src/physical_plan/expressions.rs:1404]: good question; the code blocks are not exactly the same, there is a small difference; notice how in `(ColumnarValue::Array(array), ColumnarValue::Scalar(scalar))` we have `Operator::Lt => binary_array_op_scalar!(array, scalar.clone(), lt)`, but under `(ColumnarValue::Scalar(scalar), ColumnarValue::Array(array))` we have `Operator::Lt => binary_array_op_scalar!(array, scalar.clone(), gt)`;
this is because there is only one version of arrow comparison kernel functions for scalar comparison where the scalar value can only be on one side of the comparison, for example `pub fn lt_scalar<T>(left: &PrimitiveArray<T>, right: T::Native) -> Result<BooleanArray>`

File: rust/datafusion/src/physical_plan/mod.rs
Comment (yordan-pavlov) [reply rust/datafusion/src/physical_plan/mod.rs:104]: possibly; where / how could you see the `ColumnarValue` enum used in core arrow? also wouldn't `ScalarValue` need to move as well?

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (yordan-pavlov) [reply rust/datafusion/src/physical_plan/expressions.rs:1007]: this code uses the `paste::expr!` macro, similar to how it's used in the existing `macro_rules! compute_utf8_op` to generate the scalar function name, such as `lt_scalar`, from the `$OP` parameter (which could have a value of `lt`) and the suffix `_scalar`

File: rust/datafusion/src/physical_plan/expressions.rs
Comment (andygrove) [reply rust/datafusion/src/physical_plan/expressions.rs:1404]: Ok, that makes sense. I hadn't looked closely enough to see the differences.

File: rust/datafusion/src/physical_plan/mod.rs
Comment (nevi-me) [reply rust/datafusion/src/physical_plan/mod.rs:104]: C++ uses a `Datum`, which is also an enum over Scalar, Array, and a few other things.
We could have a separate module called `arrow::scalar`, then in the long run we could convert the compute kernels to take `Datum`, and push the optimisations of "array vs scalar" there

Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/mod.rs:104]: I agree.

It would also simplify the API of many vertical operations (e.g. aggregates), as their input and result would have a common type. I have a branch on which I am doing that, but I have not finish it yet.

File: rust/datafusion/src/physical_plan/filter.rs
Comment (jorgecarleitao) [inline rust/datafusion/src/physical_plan/filter.rs:27]: This also improves performance because it avoids computing a `FilterContext` per column.

File: rust/datafusion/src/physical_plan/mod.rs
Comment (nevi-me) [reply rust/datafusion/src/physical_plan/mod.rs:104]: I also went down that rabbit hole over a year ago, yeah it's not easy

**Review by alamb — APPROVED** [review]
I love this PR -- thanks @yordan-pavlov  -- I think we should merge this once we get this rebased and the tests are passing. Let me know if you need any help -- this optimization is directly relevant to work we are doing in my work project. 

I am going to try and fire up my TPCH benchmark locally and see if I can get any more performance data

**Comment by alamb** [comment]
I tried to run this code against our simple TPCH Q1 implementation (which is dominated by evaluating expressions against constants), and sadly it hit an error

```
cd arrow/rust/benchmarks
cargo run --release --bin tpch -- --iterations 3 --path /Users/alamb/Software/tpch_data/SF10-parquet-64 --format parquet --query 1 --batch-size 4096
    Finished release [optimized] target(s) in 0.13s
     Running `/Users/alamb/Software/arrow2/rust/target/release/tpch --iterations 3 --path /Users/alamb/Software/tpch_data/SF10-parquet-64 --format parquet --query 1 --batch-size 4096`
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 2, batch_size: 4096, path: "/Users/alamb/Software/tpch_data/SF10-parquet-64", file_format: "parquet" }

Error: ArrowError(ExternalError(General("Scalar values on left side of operator - are not supported")))
```

I plan to figure out what is going on and try and fix it tomorrow morning, US Eastern time. As I said this I think this PR is a major step forward and I want to see it merged!

Andrew

**Comment by andygrove** [comment]
I would recommend that we implement an optimizer rule to "swap" the order of expressions when we see an unsupported combination such as a scalar on the left. The logic for this rule already exists in this PR. The benefit of having it as an optimizer rule is that it would handle invalid plans created from both the SQL and DataFrame APIs.

**Comment by yordan-pavlov** [comment]
thanks for all the comments, I will try to rebase this evening

**Comment by yordan-pavlov** [comment]
I have now rebased the branch on the latest changes from master; will try to review comments tomorrow.

**Comment by yordan-pavlov** [comment]
> I tried to run this code against our simple TPCH Q1 implementation (which is dominated by evaluating expressions against constants), and sadly it hit an error

@alamb it might be necessary to fallback to generating an array where the scalar value is repeated, for some operations that do not have a version which accepts a scalar argument

File: rust/datafusion/src/physical_plan/mod.rs
Comment (yordan-pavlov) [reply rust/datafusion/src/physical_plan/mod.rs:114]: I agree with @alamb ; I think if we wanted to move `ColumnarValue`  into `arrow` it would be better to do that in a separate PR after this one has been merged

Comment (wesm) [reply rust/datafusion/src/physical_plan/mod.rs:114]: Note that our use of `Datum` in C++ is **far from perfect**, it's actually a relatively complex data structure and I've contemplated using something simplified that doesn't have as many non-trivial C++ objects inside it in the internals of function execution

**Comment by alamb** [comment]
> @alamb it might be necessary to fallback to generating an array where the scalar value is repeated, for some operations that do not have a version which accepts a scalar argument

@yordan-pavlov  I agree with this plan. While long term I would like to see the ability to evaluate *every* function with constant arguments, that is definitely beyond the scope of this PR, and so keeping the fall back to constant arrays is a good idea. 

Given how important this functionality is, I suggest we focus on getting this PR into master as soon as possible and then iterating on additional functionality like additional function support, converting to canonical `expr _op_ colref` form, constant folding, hoisting ColumValue into arrow, etc. There is so much great work to do!

**Comment by yordan-pavlov** [comment]
@alamb I have now implemented falling back to scalar arrays for operations where scalar arguments are currently not supported; this should now work for operators such as "-" where previously an error was returned ("Scalar values on left side of operator - are not supported"); I have also added an extra test for this; 
could you try running the TPCH test again?

**Comment by alamb** [comment]
@yordan-pavlov  When I ran the benchmark locally again on my laptop:

```
cargo run --release --bin tpch -- --iterations 3 --path /Users/alamb/Software/tpch_data/SF10-parquet-64 --format parquet --query 1 --batch-size 4096
```

benchmarks on this branch (fd77fd6a4ee4705f8d25436c6a8dee3beb834941)
```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 2, batch_size: 4096, path: "/Users/alamb/Software/tpch_data/SF10-parquet-64", file_format: "parquet", mem_table: false }
Query 1 iteration 0 took 5371 ms
Query 1 iteration 1 took 6121 ms
Query 1 iteration 2 took 6160 ms
alamb@ip-192-168-0-133 benchmarks %
```

benchmark on master oe5fce7f6a70c370c758b133f025444c98cdd305d

```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 2, batch_size: 4096, path: "/Users/alamb/Software/tpch_data/SF10-parquet-64", file_format: "parquet", mem_table: false }
Query 1 iteration 0 took 7640 ms
Query 1 iteration 1 took 7808 ms
Query 1 iteration 2 took 7765 ms
```

So it seems to me the the performance improved a bit. I am second guessing my performance setup here - it is with a laptop and I do wonder if my CPU is being throttled.

Regardless, I think this is good enough results!

**Comment by yordan-pavlov** [comment]
@alamb thanks for running that test again; even if comparisons against scalar values are significantly faster, there is much more to the tests, such as loading data, etc.; I will be looking at that next. Overall though, good result as you said.

File: rust/datafusion/src/physical_plan/filter.rs
Comment (yordan-pavlov) [reply rust/datafusion/src/physical_plan/filter.rs:30]: @jorgecarleitao that's exactly right, good spot

**Review by jorgecarleitao — APPROVED** [review]
LGTM! Super cool!

**Pull request closed without merging**
